sphinx/ext/autodoc/__init__.py

```diff
@@ -1198,22 +1206,72 @@ def format_args(self, **kwargs: Any) -> str:
-        # __init__ written in C?
-        if initmeth is None or \
-                inspect.is_builtin_class_method(self.object, '__init__') or \
-                not(inspect.ismethod(initmeth) or inspect.isfunction(initmeth)):
-            return None
-        try:
-            self.env.app.emit('autodoc-before-process-signature', initmeth, True)
-            sig = inspect.signature(initmeth, bound_method=True)
+        def get_user_defined_function_or_method(obj: Any, attr: str) -> Any:
+            """ Get the `attr` function or method from `obj`, if it is user-defined. """
+            if inspect.is_builtin_class_method(obj, attr):
+                return None
+            attr = self.get_attr(obj, attr, None)
+            if not (inspect.ismethod(attr) or inspect.isfunction(attr)):
+                return None
+            return attr
+
+        sig = None
+
+        # this sequence is copied from inspect._signature_from_callable
+
+        # First, let's see if it has an overloaded __call__ defined
+        # in its metaclass
+        if sig is None:
+            call = get_user_defined_function_or_method(type(self.object), '__call__')
+
+            if call is not None:
+                if "{0.__module__}.{0.__qualname__}".format(call) in _METACLASS_CALL_BLACKLIST:
+                    call = None
+
+            if call is not None:
+                self.env.app.emit('autodoc-before-process-signature', call, True)
+                try:
+                    sig = inspect.signature(call, bound_method=True)
+                except TypeError:
+                    pass
+
+        # Now we check if the 'obj' class has a '__new__' method
+        if sig is None:
+            new = get_user_defined_function_or_method(self.object, '__new__')
+            if new is not None:
+                self.env.app.emit('autodoc-before-process-signature', new, True)
+                try:
+                    sig = inspect.signature(new, bound_method=True)
+                except TypeError:
+                    pass
+
+        # Finally, we should have at least __init__ implemented
+        if sig is None:
+            init = get_user_defined_function_or_method(self.object, '__init__')
+            if init is not None:
+                self.env.app.emit('autodoc-before-process-signature', init, True)
+                try:
+                    sig = inspect.signature(init, bound_method=True)
+                except TypeError:
+                    pass
+
+        # None of the attributes are user-defined, so fall back to let inspect
+        # handle it.
+        if sig is None:
+            # We don't know the exact method that inspect.signature will read
+            # the signature from, so just pass the object itself to our hook.
+            self.env.app.emit('autodoc-before-process-signature', self.object, False)
+            try:
+                sig = inspect.signature(self.object, bound_method=False)
+            except TypeError:
```

Comment:
It seems `inspect.signature()` raises `ValueError` also. Please handle it too.
```
Traceback (most recent call last):
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 395, in format_signature
    args = self.format_args(**kwargs)
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 1266, in format_args
    sig = inspect.signature(self.object, bound_method=False)
  File "/Users/tkomiya/work/sphinx/sphinx/util/inspect.py", line 406, in signature
    signature = inspect.signature(subject)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 3093, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2842, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2372, in _signature_from_callable
    raise ValueError(
ValueError: no signature found for builtin type <class 'sphinx.errors.SphinxError'>
```

Reply:
A testcase for exception class is also needed:
```
$ python
Python 3.8.2 (default, Mar  2 2020, 00:44:41)
[Clang 11.0.0 (clang-1100.0.33.17)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import inspect
>>> inspect.signature(Exception)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 3093, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2842, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2372, in _signature_from_callable
    raise ValueError(
ValueError: no signature found for builtin type <class 'Exception'>
```

Reply:
Test added.